text/0000-dockerfiles.md

```diff
@@ -27,25 +27,65 @@ This RFC proposes that we replace stackpacks with multi-purpose build-time and r
 # How it Works
 [how-it-works]: #how-it-works
 
-Note: kaniko, buildah, BuildKit, the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
 
-### App-specified Dockerfiles
+### Builder-specified Dockerfiles
 
-A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory. A run.Dockerfile is applied to the selected runtime base image after the detection phase. A build.Dockerfile is applied to the build-time base image before the detection phase.
+A builder may specify any number of executable "hooks" in `/cnb/hooks.d/`.
+Hook files in this directory are executed in the context of app directory and must output Dockerfiles that are applied to the build-time and runtime base images before an image is built.
 
-Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  
+If a hook exits with a non-zero status value, the build fails. If a hook exits with a zero status value and no output, the hook is ignored. Directories and non-executable files are ignored.
+
+Each executable must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>`, where build, run, or empty specify the build-time base image, runtime base image, or both bash images, respectively. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future. 
+
+A runtime Dockerfile is applied to the selected runtime base image after the detection phase.
+A build-time Dockerfile is applied to the build-time base image before the detection phase.
+
+Both Dockerfiles must accept `base_image` and `build_id` args.
+The `base_image` arg allows the lifecycle to specify the original base image.
```

Comment:
Are we allowed to override/ignore this `base_image` and specify our own?

Reply:
Anything is valid, as long as the Dockerfile accepts the ARG and the resulting image is spec-compliant.